src/Database/Esqueleto/Experimental.hs

```diff
@@ -344,91 +462,187 @@ import GHC.TypeLits
+  toSetOperation :: a -> SqlSetOperation b
+
+instance ToSetOperation (SqlSetOperation a) a where
+  toSetOperation = id
+instance ToSetOperation (SqlQuery a) a where
+  toSetOperation = SelectQueryP Never
+instance (ToSetOperation a c, ToSetOperation b c) => ToSetOperation (Union a b) c where
+  toSetOperation (Union a b) = SqlSetUnion (toSetOperation a) (toSetOperation b)
+instance (ToSetOperation a c, ToSetOperation b c) => ToSetOperation (UnionAll a b) c where
+  toSetOperation (UnionAll a b) = SqlSetUnionAll (toSetOperation a) (toSetOperation b)
+instance (ToSetOperation a c, ToSetOperation b c) => ToSetOperation (Except a b) c where
+  toSetOperation (Except a b) = SqlSetExcept (toSetOperation a) (toSetOperation b)
+instance (ToSetOperation a c, ToSetOperation b c) => ToSetOperation (Intersect a b) c where
+  toSetOperation (Intersect a b) = SqlSetIntersect (toSetOperation a) (toSetOperation b)
+
+type family SetOperationT a where
+  SetOperationT (Union a b) = SetOperationT a
+  SetOperationT (UnionAll a b) = SetOperationT a
+  SetOperationT (Except a b) = SetOperationT a
+  SetOperationT (Intersect a b) = SetOperationT a
+  SetOperationT (SqlQuery a) = a
+  SetOperationT (SqlSetOperation a) = a
+
+{-# DEPRECATED SelectQuery "/Since: 3.4.0.0/ - \
+    It is no longer necessary to tag 'SqlQuery' values with @SelectQuery@" #-}
 pattern SelectQuery :: SqlQuery a -> SqlSetOperation a
 pattern SelectQuery q = SelectQueryP Never q
 
+
 -- | Data type that represents the syntax of a 'JOIN' tree. In practice,
 -- only the @Table@ constructor is used directly when writing queries. For example,
 --
 -- @
 -- select $ from $ Table \@People
 -- @
 data From a where
-  Table         :: PersistEntity ent => From (SqlExpr (Entity ent))
-  SubQuery      :: (SqlSelect a' r, SqlSelect a'' r', ToAlias a, a' ~ ToAliasT a, ToAliasReference a', ToAliasReferenceT a' ~ a'')
-                => SqlQuery a
-                -> From a''
-  SqlSetOperation :: (SqlSelect a' r, ToAlias a, a' ~ ToAliasT a, ToAliasReference a', ToAliasReferenceT a' ~ a'')
-                  => SqlSetOperation a
-                  -> From a''
-  InnerJoinFrom :: From a
-                -> (From b, (a :& b) -> SqlExpr (Value Bool))
-                -> From (a :& b)
-  CrossJoinFrom :: From a
-                -> From b
-                -> From (a :& b)
-  LeftJoinFrom  :: ToMaybe b
-                => From a
-                -> (From b, (a :& ToMaybeT b) -> SqlExpr (Value Bool))
-                -> From (a :& ToMaybeT b)
-  RightJoinFrom :: ToMaybe a
-                => From a
-                -> (From b, (ToMaybeT a :& b) -> SqlExpr (Value Bool))
-                -> From (ToMaybeT a :& b)
-  FullJoinFrom  :: (ToMaybe a, ToMaybe b )
-                => From a
-                -> (From b, (ToMaybeT a :& ToMaybeT b) -> SqlExpr (Value Bool))
-                -> From (ToMaybeT a :& ToMaybeT b)
+  Table                 :: PersistEntity ent => From (SqlExpr (Entity ent))
```

Comment:
Formatting suggestion: instead of right-aligning on the longest constructor name, indent the type sig further

eg

```haskell
data From a where
  Table 
    :: PersistEntity ent => From (SqlExpr (Entity ent))
  SubQuery
    :: ( constraints )
    => SqlQuery a
    -> From a''
    -> From a
-- etc
```

This will make for better diffs in the future.